### DISCUSSION TITLE
Migration from v6 to v7 form data collection behavior

### DISCUSSION THREAD

**Discussion by plamenkoyovchev** [react-hook-form/react-hook-form#7780] [Q&A] [answered]
Hello, 

I've prepared little example of my issue, here it is.

https://codesandbox.io/s/react-hook-form-get-started-forked-rcee6?file=/src/index.js

I have conditional rendering of my last input field. As you can see if the checkbox is checked I have both input fields as props of my data object: nickname and secondNickname. I'm expecting only the visible element to be included in the data object.

In v6 if element is not visible in the form it was automatically omitted and was not part of data object.

How can I achieve this behavior in v7 ?

**Comment by bluebill1049** [comment, marked as answer]
Set shouldUnegister true

**Reply by plamenkoyovchev** [reply]
I have nested form input names: "details.recurring.date", when I set shouldUnregister to true this sets `date` to null which is partially ok but I need also `recurring` object to be null.

Should I do something like `unregister("details.recurring")` when conditional rendering initiator changes ?

**Reply by bluebill1049** [reply]
yes.

**Reply by plamenkoyovchev** [reply]
`const jobType = useWatch({
		name: jobTypeFormName
	});
`

```
useEffect(() => {
if (jobType) {
	if (jobType === jobTypes.oneTime.value) {
		unregister(recurringFormPrefix);
	} else {
		unregister(oneTimeFormPrefix);
	}
}
}, [jobType, oneTimeFormPrefix, recurringFormPrefix, unregister]);
```

After calling `unregister(recurringFormPrefix);` form data still has the recurring object on form submit.
Look:

![image](https://user-images.githubusercontent.com/3595159/153201490-7db2e0d7-074c-4cff-b88c-92fcdf0e725b.png)

Any idea what I'm doing wrong ?

**Reply by bluebill1049** [reply]
can you double-check if the `register` gets invoked again?

**Reply by plamenkoyovchev** [reply]
I am using Controller component for all my FormInput components and shouldUnregister={true}. I've never called register.

**Reply by Moshyfawn** [reply]
Here's the intended behavior: take a look at [this CSB](https://codesandbox.io/s/loving-andras-xmscb?file=/src/styles.css)

**Reply by plamenkoyovchev** [reply]
Aaaaaah! Found it! I should've just set shouldUnregister: true in my FormContainer component. Thank you guys for the help!

```
const methods = useForm({
		mode: validationMode,
		reValidateMode: validationMode,
		shouldUnregister: true
	});
```
